### ISSUE TITLE
No log driver named 'gelf' is registered

### ISSUE THREAD

**Issue by mle-ii** [docker/for-win#1143]
### Expected behavior
Specifying gelf as the log-driver should use that plugin for sending logs to gelf and should then send log data to the IP:PORT specified.

### Actual behavior
Both using the option in the docker run command and via the daemon.json config file don't work and report something like the following:

> C:\Program Files\Docker\Docker\Resources\bin\docker.exe: Error response from daemon: logger: no log driver named 'gelf' is registered.
> See 'C:\Program Files\Docker\Docker\Resources\bin\docker.exe run --help'.

### Information
```
Version: 17.09.0-ce-win31 (13514)
Channel: edge
Sha1: ff8d09a22918de4afe37795cbcaadd1c4da5cd9c
Started on: 2017/09/29 11:16:24.755
Resources: C:\Program Files\Docker\Docker\Resources
OS: Windows 10 Pro
Edition: Professional
Id: 1703
Build: 15063
BuildLabName: 15063.0.amd64fre.rs2_release.170317-1834
```

The docs seem to imply that this driver is supported on Windows and I've seen various folks report their docker info logs that seem to indicate that they have it as well.
https://docs.docker.com/engine/admin/logging/overview/#supported-logging-drivers
https://docs.docker.com/engine/admin/logging/gelf/#usage

From `docker info` it is not listed.
```
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
```

### Steps to reproduce the behavior

  1.  `docker run -i --log-driver=gelf --log-opt gelf-address=udp://127.0.0.1:12345 microsoft/powershell:nanoserver powershell`

**Comment by friism** [comment]
It looks like the gelf driver is not compiled for windows and so is not supported there: https://github.com/moby/moby/blob/master/daemon/logger/gelf/gelf.go#L1

**Comment by mle-ii** [comment]
@friism So I need to open an issue there on the Moby github to get Docker for Windows to support gelf?  Or are you saying that Docker for Windows doesn't plan to support gelf?

Sorry, new to reporting issues that seemed to be Windows specific and thought they would go here.

We use graylog to do our Windows logging and this makes getting some of the logging for containers a bit more problematic.

**Comment by friism** [comment]
gelf will work when running Linux containers on Windows with Docker for Windows, but not when running Docker Windows containers (sorry, I know this is a little confusing).

You could try removing the compiler directive from here: https://github.com/moby/moby/blob/master/daemon/logger/gelf/gelf.go#L1

And also remove this file: https://github.com/moby/moby/blob/master/daemon/logger/gelf/gelf_unsupported.go

... then build your own `dockerd.exe` daemon, run it and see if you can get it logging to gelf.

Instructions: https://github.com/moby/moby/blob/master/Dockerfile.windows#L11

**Comment by mle-ii** [comment]
@friism Ok, I got it to "work" by also adding this line:
https://github.com/moby/moby/blob/master/daemon/logdrivers_linux.go#L9

To this file after this line:
https://github.com/moby/moby/blob/master/daemon/logdrivers_windows.go#L8

I say "work" as I need to figure out the formatting for the gelf that it sends as by default it appears our graylog server doesn't handle the format that's being sent.  Likely just a config on our graylog instance.  But with the changes you referred to along with the additional change I mentioned it's sending json blobs via the gelf logger which is a step in the right direction.

So where do we go from here?  I'm not convinced I understand the docker daemon enough to know what else might need to change on that side of things.  I can raise the issue there with a report with these further details if need be.  

This isn't currently a blocker for us as we're still working on getting our environment/infrastructure figured out and set up, the bigger blocker will be the UDP messages being dropped in the other issue I reported.  Right now we're in the analysis stage to see what issues there may or may not be.  Thanks for your help in narrowing down where the issue might be.
